### PR TITLE
Dynamo remaps legacy allgather to traceable one

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102232

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: test/distributed/test_inductor_collectives.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import torch
 from torch._C import FileCheck
 # for some reason importing functional collectives after dynamo breaks collectives handling!
+# import torch.distributed
```
Comment (voznesenskym) [inline test/distributed/test_inductor_collectives.py:8]: nit rm

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -440,6 +440,31 @@ def index_source(key):
...
+                reduce_scatter_tensor as legacy_reducescatter,
+            )
+
+            # TODO assert that async_op is false, and, possibly some other limitations
+            # current remapping is very brittle, in regards to handling args at call time
+            # and seriously.. can't import properly, and also getting line-too-long linter
+            traceable_collective_remaps = {
+                legacy_allgather: all_gather_tensor_inplace,
+                legacy_reducescatter: reduce_scatter_tensor_inplace,
+            }
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:462]: If you build this remaps outside the if/elif chain here, you can remove most of the function here

```
traceable_collective_remaps = {
torch.distributed.distributed_c10d.all_gather_tensor_inplace : torch.distributed._functional_collectives.all_gather_tensor_inplace, 
torch.distributed.distributed_c10d.reduce_scatter_tensor_inplace : torch.distributed._functional_collectives.reduce_scatter_tensor_inplace, 
}
```

```
elif value in traceable_collective_remaps:
    return UserFunctionVariable(
                traceable_collective_remaps[value],
                source=self.source,
                guards=make_guards(GuardBuilder.FUNCTION_MATCH),
            )
```

File: test/distributed/test_inductor_collectives.py [hunk 3]
```diff
@@ -375,6 +376,57 @@ def func(inp, *, pg):
...
+        outputs = torch.empty(global_size, device=self.device)
+        correct_outputs = torch.empty(global_size, device=self.device)
+        counter = CompileCounter()
+        compiled = torch.compile(func, backend=counter, fullgraph=True)
+        compiled(inputs, outputs, pg=GroupMember.WORLD)
+        func(inputs, correct_outputs, pg=GroupMember.WORLD)
+        assert counter.frame_count == 1
+
+        # should test more precisely, but the 3 is supposed to be (all_gather, wait, copy_)
+        assert counter.op_count == 3
```
Comment (voznesenskym) [inline test/distributed/test_inductor_collectives.py:403]: self.assertEqual in tests

File: test/distributed/test_inductor_collectives.py [hunk 4]
```diff
@@ -375,6 +376,57 @@ def func(inp, *, pg):
         assert counter.op_count == 2
         assert same(out, correct)
 
+    def test_dynamo_rewrite_dist_all_gather(self):
```
Comment (voznesenskym) [inline test/distributed/test_inductor_collectives.py:379]: How does this work w/o init_process_group?

File: torch/_dynamo/variables/builder.py [hunk 5]
```diff
@@ -440,6 +440,31 @@ def index_source(key):
                 guards=make_guards(GuardBuilder.FUNCTION_MATCH),
             )
         # NB: These can't be put in type_dispatch, they have to run later
+        elif value in {
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:443]: also assumption of hashable in this tree is unsafe, can't just `value in` - Please add a check here that this is a function, or do the comparison directly, or make a list.

File: torch/distributed/_functional_collectives.py [hunk 6]
```diff
@@ -517,3 +517,31 @@ def _register_ops():
...
+def all_gather_tensor_inplace(
+    output: torch.Tensor,
+    input: torch.Tensor,
+    group,  # TODO add a type,
+    async_op: bool = False,
+    tag: str = "",
+    gather_dim: int = 0
+):
+    assert not async_op, "Can't remap async version of inplace op to functional collective"
+    return output.copy_(all_gather_tensor(input, gather_dim, group, tag))
```
Comment (voznesenskym) [inline torch/distributed/_functional_collectives.py:535]: I don't really understand how this fixes the need to define how to trace PG?

File: torch/_dynamo/variables/builder.py (hunk 5 above)
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:443]: You also have to do this before the `is_allowed` check or allowlist shuffling will short circuit past this.

File: torch/_dynamo/variables/builder.py [hunk 7]
```diff
@@ -440,6 +440,31 @@ def index_source(key):
...
+            )
+
+            # TODO assert that async_op is false, and, possibly some other limitations
+            # current remapping is very brittle, in regards to handling args at call time
+            # and seriously.. can't import properly, and also getting line-too-long linter
+            traceable_collective_remaps = {
+                legacy_allgather: all_gather_tensor_inplace,
+                legacy_reducescatter: reduce_scatter_tensor_inplace,
+            }
+            return UserFunctionVariable(
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:463]: Does this inline through it? Can you dump the fx graph? I am not sure this is what we want...

Comment (wconstab) [reply torch/_dynamo/variables/builder.py:463]: yes, this inlines through it.

It's good bc i wrote it as a user func- it calls first `copy_` and then `all_gather_blah` both of which are traceable.  

I looked at the fx graph from running my unit test with `TORCH_LOGS=graph_code` or whatever the right name is.  IT showed 3 ops - all_gather, wait, copy_.

File: torch/_dynamo/variables/builder.py (hunk 5 above)
Comment (wconstab) [reply torch/_dynamo/variables/builder.py:443]: this part is going over my head.

How come it works for my unit test and not real fsdp?

File: torch/distributed/_functional_collectives.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/_functional_collectives.py:535]: PG comes into this function as a UserDefinedObjectVariable (iirc.  or whatever).

inside all_gather_tensor, it gets used in `expand_group` and is used as a key in a dict.  From there it turns into ints and a string, which are new constants we use in the actual traceable ops.

File: torch/_dynamo/variables/builder.py (hunk 5 above)
Comment (wconstab) [reply torch/_dynamo/variables/builder.py:443]: yep good point. i will do this part another way

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (wconstab) [reply torch/_dynamo/variables/builder.py:462]: yea. i intend to put this dict in _functional_collectives.py, and then use it for the elif check as well as the mapping.

but i ran into circular import hell.

i didn't figure that part out yet but ill get there

File: torch/distributed/_functional_collectives.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/_functional_collectives.py:535]: Ok, we ended up adding ProcesGroupVariable to handle other cases (e.g. pg.rank() .size() calls), but PGVariable still isn't a critical part of how this rewrite part works.

**Review by voznesenskym — APPROVED** [review]

**Review by voznesenskym — APPROVED** [review]

**Review by voznesenskym — APPROVED** [review]
